samples/net/openthread/coap_client/src/coap_utils.c

```diff
@@ -0,0 +1,262 @@
+#include "coap_utils.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(coap_utils, LOG_LEVEL_DBG);
+
+#include <errno.h>
+#include <sys/printk.h>
+#include <sys/byteorder.h>
+#include <zephyr.h>
+
+#include <net/socket.h>
+#include <net/net_mgmt.h>
+#include <net/net_ip.h>
+#include <net/udp.h>
+#include <net/coap.h>
+
+#define PEER_PORT 5683
+#define MAX_COAP_MSG_LEN 256
+#define COAP_VER 1
+#define COAP_TOKEN_LEN 8
+#define COAP_MAX_REPLIES 1
+#define COAP_POOL_SLEEP 500
+#define COAP_RECEIVE_STACK_SIZE 500
+
+static K_THREAD_STACK_DEFINE(receive_stack_area, COAP_RECEIVE_STACK_SIZE);
+static struct k_thread receive_thread_data;
+
+static int sock;
+const static int nfds = 1;
+struct pollfd fds;
+struct coap_reply m_replies[COAP_MAX_REPLIES];
+
+static void coap_receive()
+{
+	static u8_t buf[MAX_COAP_MSG_LEN + 1];
+
+	struct coap_packet response;
+	struct coap_reply *reply = NULL;
+	static struct sockaddr from_addr;
+	socklen_t from_addr_len;
+	int len;
+	int ret;
+
+	while (1) {
+		if (poll(&fds, nfds, K_FOREVER) < 0) {
+			LOG_ERR("Error in poll:%d", errno);
+			errno = 0;
+			k_sleep(COAP_POOL_SLEEP);
+			continue;
+		}
+
+		if (fds.revents & POLLERR) {
+			LOG_ERR("Error in poll.. waiting a moment.");
+			fds.revents = 0;
+			k_sleep(COAP_POOL_SLEEP);
+			continue;
+		}
+
+		if (fds.revents & POLLHUP) {
+			LOG_ERR("Error in poll: POLLHUP");
+			fds.revents = 0;
+			continue;
+		}
+
+		if (fds.revents & POLLNVAL) {
+			LOG_ERR("Error in poll: POLLNVAL - fd not open");
```

Comment:
Shall we re-open the socket then? In general, I think that's it's wise to close and reopen in case of error - I wouldn't trust the underlying implementation that the socket was left in a reliable state.

Reply:
changed